api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java

```diff
@@ -329,5 +329,36 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
       return ROWS_MIGHT_MATCH;
     }
+
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      int pos = Accessors.toPosition(ref.accessor());
+      PartitionFieldSummary fieldStats = stats.get(pos);
+
+      // values are all null (and stats exist) and literal cannot contain null
+      if (fieldStats.lowerBound() == null) {
+        return ROWS_CANNOT_MATCH;
+      }
+
+      ByteBuffer prefixAsBytes = lit.toByteBuffer();
+
+      Comparator<ByteBuffer> comparator = Comparators.unsignedBytes();
+
+      ByteBuffer lower = fieldStats.lowerBound();
+      // truncate lower bound so that its length in bytes is not greater than the length of prefix
+      int lowerLength = Math.min(prefixAsBytes.remaining(), lower.remaining());
+      int lowerCmp = comparator.compare(BinaryUtil.truncateBinary(lower, lowerLength), prefixAsBytes);
+      if (lowerCmp == 0) {
+        ByteBuffer upper = fieldStats.upperBound();
+        // truncate upper bound so that its length in bytes is not greater than the length of prefix
+        int upperLength = Math.min(prefixAsBytes.remaining(), upper.remaining());
+        int upperCmp = comparator.compare(BinaryUtil.truncateBinary(upper, upperLength), prefixAsBytes);
+        if (upperCmp == 0) {
+          return ROWS_CANNOT_MATCH;
+        }
+      }
```

Comment:
I had originally made a utility function, `BinaryUtils.startsWith(ByteBuffer value, ByteBuffer prefix, Comparator<ByteBuffer> cmp)`, however there are a number of assumptions that are valid once we've gotten to the evaluators that don't necessarily hold in the general case, e.g. that the ByteBuffers are at `position` 0 (which should be the case here), so I decided to inline the definition the handful of times that it's needed as it's not too complicated.

If we think the utility function would be cleaner, I'd be happy to add it back in and also update STARTS_WITH to use it as well (it's inlined in the same manner, with different logic on the comparison result).

Reply:
Sounds good to me. Could you make the same changes as in the `InclusiveMetricsEvaluator`?